server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -525,13 +524,12 @@ public DirectoryReader getDirectoryReader() {
 
     private static class MutableQueryTimeout implements ExitableDirectoryReader.QueryCancellation {
 
-        private final Set<Runnable> runnables = new HashSet<>();
+        private final List<Runnable> runnables = new ArrayList<>();
```

Comment:
:+1: the default size of 10 makes sense here, if anything I expect it's a little oversized. Might be worth a comment to that effect tho.